[AIRFLOW-1345] Dont expire TIs on each iteration

saguziel opened the pull request:
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1345


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes: I removed the commit on each iteration of the loop. At the beginning of this function, we batch query TIs. The commit expires all these TIs, forcing us to requery them on each iteration. This is very slow.


### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason: Doesn't change existing behavior. There's already tests in SchedulerJob


### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"


@aoen @bolkedebruin

aoen commented on airflow/jobs.py:
```diff
@@ -1140,6 +1138,7 @@ def _execute_task_instances(self,
 
                 open_slots -= 1
                 dag_id_to_possibly_running_task_count[dag_id] += 1
+        session.commit()
```
This will have the consequence of preventing any tasks from being executed if there is an issue in e.g. the executor that throws an exception (like we saw with the CeleryExecutor at Airbnb). We could commit the session in a finally block, but it's probably fine the way it is.

aoen commented on airflow/jobs.py:
```diff
@@ -1087,59 +1090,61 @@ def _execute_task_instances(self,
+            # Set the state to queued
+            task_instance.refresh_from_db(lock_for_update=True, session=session)
+            if task_instance.state not in states:
+                self.logger.info("Task {} was set to {} outside this scheduler."
```
Previously this continue and not increment open_slots and now it will (so less tasks than should be will get sent to the executor). Unfortunately this method is pretty complex, I think this new change might warrant some additional testing (and eventually some modularization but that is probably out of scope for this PR).

saguziel replied:
@aoen  I don't think this is a breaking change. It will still pick up more from the pool on the next iteration.

aoen replied:
It's a relatively minor bug, but it's still a bug.

saguziel replied:
Kind of debatable IMO. There's no contract to get the maximum number of tasks queued in the presence of race conditions.

saguziel replied:
Anyway it is minor enough.

bolkedebruin commented on airflow/jobs.py:
```diff
@@ -1087,59 +1090,61 @@ def _execute_task_instances(self,
-                                             if not task_instance.queued_dttm
-                                             else task_instance.queued_dttm)
-                session.merge(task_instance)
+        for task_instance in tis_to_queue:
```
You could do this in one db query so you don't need do it per TI. 

pseudo code:

select TIs from TaskInstances where ti in tis_to_queue and state in states with lock for update

for ti in TIs:
  do you stuff

commit()

bolkedebruin replied:
downside is obviously a longer lock per ti

aoen commented on airflow/jobs.py:
```diff
@@ -1083,63 +1083,165 @@ def _execute_task_instances(self,
+            .filter(or_(*filter_for_ti_enqueue))
+            .all())
+
+        task_instance_str = "\n\t".join(
```
task_instance str is not used below, found this in landscape errors, might want to look through the rest of them too

saguziel replied:
fixed

saguziel committed:
[AIRFLOW-1345] Dont expire TIs on each scheduler loop

TIs get expired on commit, which causes any access to their properties
to cause a new query to the DB to be issued, causing an n+1 query issue,
even when the TI is not scheduled. This change makes all queries
batches, which will make scheduling substantially faster.
